### PR TITLE
more v0.2.0 spec updates

### PR DESCRIPTION
**Pull request by tersec**

### CODE REVIEW

File: beacon_chain/spec/beaconstate.nim [hunk 1]
```diff
@@ -249,6 +254,7 @@ func get_attestation_participants*(state: BeaconState,
   assert verify_bitfield(bitfield, len(crosslink_committee))
 
   # Find the participating attesters in the committee
+  # TODO investigate functional library / approach to help avoid loop bugs
```
Comment (mratsim) [inline beacon_chain/spec/beaconstate.nim]: How would you want it to look like in pseudocode?

File: beacon_chain/spec/datatypes.nim [hunk 2]
```diff
@@ -166,8 +166,14 @@ const
...
 type
   ValidatorIndex* = range[0'u32 .. 0xFFFFFF'u32] # TODO: wrap-around
+
+  ## In principle, these would be better as distinct types. That's a good
+  ## TODO eventually, but Danny has confirmed that the SSZ types will use
+  ## primitive (uint64, etc) types and helper functions annotated ones so
+  ## it would just create pointless casts for now.
   SlotNumber* = uint64
   EpochNumber* = uint64
+  Gwei* = uint64
```
Comment (mratsim) [inline beacon_chain/spec/datatypes.nim]: seems unrelated to the PR?

File: beacon_chain/spec/validator.nim [hunk 3]
```diff
@@ -12,33 +12,101 @@ import
...
-  result = seed
-  for i in 0 ..< 8:
-    result.data[31 - i] = result.data[31 - i] xor byte((x shr i*8) and 0xff)
-
-# https://github.com/ethereum/eth2.0-specs/blob/v0.1/specs/core/0_beacon-chain.md#get_shuffling
+# https://github.com/ethereum/eth2.0-specs/blob/v0.2.0/specs/core/0_beacon-chain.md#bytes_to_int
+func bytes_to_int(data: seq[byte]): uint64 =
+  assert data.len == 8
+
+  # Little-endian
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: static: assert cpuEndian == LittleEndian, "This code is only compatible with little endian platforms."

File: beacon_chain/spec/validator.nim [hunk 4]
```diff
@@ -12,33 +12,101 @@ import
...
+  result = 0
+  for i in countdown(7, 0):
+    result = result * 256 + data[i]
+
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: the whole thing should be in helpers/utils/tools/misc

File: beacon_chain/spec/validator.nim [hunk 5]
```diff
@@ -12,33 +12,101 @@ import
...
+func int_to_bytes4(x: uint64): array[4, byte] =
+  assert x >= 0'u64
+  assert x < 2'u64^32
+
+  # Little-endian
+  result[0] = ((x shr  0) and 0xff).byte
+  result[1] = ((x shr  8) and 0xff).byte
+  result[2] = ((x shr 16) and 0xff).byte
+  result[3] = ((x shr 24) and 0xff).byte
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: Same, assert + move to helpers

File: beacon_chain/spec/validator.nim [hunk 6]
```diff
@@ -12,33 +12,101 @@ import
...
+
+# https://github.com/ethereum/eth2.0-specs/blob/v0.2.0/specs/core/0_beacon-chain.md#int_to_bytes1-int_to_bytes2-
+# Borderline worth metaprogramming these, since in practice, only a couple are used.
+func int_to_bytes1(x: int): array[1, byte] =
+  assert x >= 0
+  assert x < 256
+
+  result[0] = x.byte
+
+func int_to_bytes4(x: uint64): array[4, byte] =
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: it probably should be renamed later to `u64_to_byte4`

when reading the code that depended on it below I forgot the type signature
and was checking if we could end up with negative values:

```Nim
    let
      pivot = bytes_to_int(eth2hash(pivot_buffer).data[0..7]) mod list_size
      flip = (pivot - index) mod list_size
```

File: beacon_chain/spec/validator.nim [hunk 7]
```diff
@@ -12,33 +12,101 @@ import
...
+  var source_buffer: array[(32+1+4), byte]
+
+  for round in 0 ..< SHUFFLE_ROUND_COUNT:
+    pivot_buffer[0..31] = seed.data
+    let round_bytes1 = int_to_bytes1(round)[0]
+    pivot_buffer[32] = round_bytes1
+
+    let
+      pivot = bytes_to_int(eth2hash(pivot_buffer).data[0..7]) mod list_size
+      flip = (pivot - index) mod list_size
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: underflow? We should probably add an assert.

File: beacon_chain/spec/validator.nim [hunk 8]
```diff
@@ -12,33 +12,101 @@ import
...
+    source_buffer[0..31] = seed.data
+    source_buffer[32] = round_bytes1
+    source_buffer[33..36] = int_to_bytes4(position div 256)
+
+    let
+      source = eth2hash(source_buffer).data
+      byte_value = source[(position mod 256) div 8]
+      bit = (byte_value shr (position mod 8)) mod 2
+
+    result = if bit != 0: flip else: result
```
Comment (mratsim) [inline beacon_chain/spec/validator.nim]: nitpick:

```Nim
if bit != 0:
  result = flip
```

would generate less C code and may be faster (or have no difference due to compiler optimisation)

File: beacon_chain/spec/datatypes.nim (hunk 2 above)
Comment (tersec) [reply beacon_chain/spec/datatypes.nim]: The PR had a misleading title; it's fixed.

File: beacon_chain/spec/validator.nim (hunk 3 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: Except that the code isn't only compatible with little-endian. It's a comment that the data representation it's working with is little-endian, per spec.

File: beacon_chain/spec/validator.nim (hunk 5 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: For assert, see response above; it's unnecessary.

File: beacon_chain/spec/validator.nim (hunk 6 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: I agree about the renaming (or at least to something like that) -- but, currently, I very much want not to relatively pointlessly change spec function names, while we're still syncing with it on an ongoing basis.

File: beacon_chain/spec/beaconstate.nim (hunk 1 above)
Comment (tersec) [reply beacon_chain/spec/beaconstate.nim]: I actually didn't have a strong opinion there. That wasn't my comment -- it just makes more sense where it is than where it was, since it was prefacing map/filter/etc approaches. Might be worth just removing though.

File: beacon_chain/spec/validator.nim (hunk 3 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: The comment's just noting that it's operating on `data` being a little-endian representation of the `uint64`, not that it only runs on little-endian architectures (I try to avoid such code, of course).

File: beacon_chain/spec/validator.nim (hunk 7 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: Underflow potential is intentional here -- it's effective supposed to operate in an abelian group mod `list size` over the `+` operation (or `-`, more or less equivalently for its purpose). https://link.springer.com/content/pdf/10.1007%2F978-3-642-32009-5_1.pdf is what the spec links -- they refer to figure 3, but figure 1 makes the intent clearer; figure 3 is just a generalized version of that:

> It is useful to be a bit more general here,working in a finite abelian group G=([N],+) instead of the group ({0,1}n,⊕) of bitstrings under xor. (For convenience, we have assumed that the group elements are named [N]={0,...,N−1}.) In this way we won’t need the number of pointsN inthe message spaceX=[N] to be a power of two—we’ll be able to encipher points on any set X=[N], just by naming a group operator, say addition modulo N. For generalizing the shuffle of Fig. 2, the value K is uniformly drawn from [N] rather than from{0,1}n, and we consider the pair of positions{X,K−X}rather than{X,K⊕X}.

The more subtle (but surprising, in a way, since they switched from their first or second shuffling to rejection sampling version precisely to avoid this) issue is that they're not uniformly drawing from [N], because of the usual foo mod list_size sort of bias issues.

File: beacon_chain/spec/validator.nim (hunk 8 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: I agree, it's better. Done.

File: beacon_chain/spec/validator.nim (hunk 4 above)
Comment (tersec) [reply beacon_chain/spec/validator.nim]: The idea was to limit the scope of each function to the minimum needed at a given spec version to facilitate static dependency/layer analysis. Still, these are actually-spec-suggested/named-helper functions, so I agree, sure, moved them to the spec/helpers.nim (where they probably most properly live).
